src/utils/generic.rs

```diff
@@ -11,28 +11,41 @@ use core::{
     },
 };
 
-use blake2::Blake2b;
+use blake2::{crypto_mac::Mac, Blake2b};
 use curve25519_dalek::scalar::Scalar;
 
 use crate::{errors::ProofError, protocols::scalar_protocol::ScalarProtocol, range_proof::RangeProof};
 
 /// Create a Blake2B deterministic nonce given a seed, label and index
-pub fn nonce(seed_nonce: &Scalar, label: &str, index: Option<usize>) -> Result<Scalar, ProofError> {
+pub fn nonce(
+    seed_nonce: &Scalar,
+    label: &str,
+    index_j: Option<usize>,
+    index_k: Option<usize>,
+) -> Result<Scalar, ProofError> {
     let encoded_label = label.as_bytes();
     if encoded_label.len() > 16 {
         // See https://www.blake2.net/blake2.pdf section 2.8
         return Err(ProofError::InvalidLength("nonce label".to_string()));
     };
-    let hasher = if let Some(salt) = index {
-        let encoded_index = salt.to_le_bytes();
-        if encoded_index.len() > 16 {
+    let mut hasher = if let Some(salt) = index_j {
+        let encoded_salt = salt.to_le_bytes();
+        if encoded_salt.len() > 16 {
             // See https://www.blake2.net/blake2.pdf section 2.8
-            return Err(ProofError::InvalidLength("nonce index".to_string()));
+            return Err(ProofError::InvalidLength("nonce index_j".to_string()));
         };
-        Blake2b::with_params(&seed_nonce.to_bytes(), &encoded_index, encoded_label)
+        Blake2b::with_params(&seed_nonce.to_bytes(), &encoded_salt, encoded_label)
     } else {
         Blake2b::with_params(&seed_nonce.to_bytes(), &[], encoded_label)
     };
+    if let Some(salt) = index_k {
+        let encoded_salt = salt.to_le_bytes();
+        if encoded_salt.len() > 16 {
+            // See https://www.blake2.net/blake2.pdf section 2.8
+            return Err(ProofError::InvalidLength("nonce index_k".to_string()));
```

Comment:
Since this index is being used as input to the hash function (and not directly as a salt parameter), this length limitation does not apply.

You could also concatenate fixed-length encodings of `j` and `k` and use this as the salt parameter; such an approach seems more canonical and in the spirit of the hash function parameter set, but it does imply further restriction on the index values.

Reply:
Good observation; I removed the length checks on both indexes as they are limited by the length of the byte representation of `usize::MAX`, which is `< 16`, and added a parameter length test. 

Interestingly enough, if the `persona` or `salt` parameters exceed 16 bytes, unhandled exceptions occur, so we have to do the length check ourselves.

Example exception passing 17 bytes into the `persona` parameter:
```
assertion failed: persona.len() <= length
thread 'utils::generic::tests::test_blake2b_parameter_lengths' panicked at 
   'assertion failed: persona.len() <= length', 
   .cargo\registry\src\github.com-1ecc6299db9ec823\blake2-0.9.2\src\blake2b.rs:4:1
```